### ISSUE TITLE
[NVFuser] Investigate modules with bad performance relative to eager

### ISSUE THREAD

**Issue by davidberard98** [pytorch/pytorch#80187]
### 🚀 The feature, motivation and pitch

In particular, `hf_DistilBert_forward_0` has ~30% regression compared to eager.

Status document (for those who have access): https://docs.google.com/document/d/1XU48ru9Cj4XFJJF6FTQ2X_QFuzyLroE1oomKGleYfP8/edit?usp=sharing

**Comment by rdspring1** [comment]
I looked into the `hf_DistilBert_forward_0` regression. In the `add + native_layer_norm` fusion, `x-mean` is calculated twice, so we perform three reductions. The `var_mean` definition computes the mean and variance separately instead of reusing the mean in the variance calculation. In the NvFuser implementation, we calculate the mean and variance together with the Welford online algorithm, so there is only one reduction.

Reference: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm

**Comment by ezyang** [comment]
So the problem is this decomp?

```
def var_mean(
    a: TensorLikeType,
    dim: Optional[DimsType] = None,
    unbiased: Optional[bool] = None,
    keepdim: bool = False,
    *,
    correction: Optional[int] = None,
):
    v = var(a, dim, unbiased, keepdim, correction=correction)
    m = mean(a, dim, keepdim)
    return v, m
```

Can we rewrite this decomp to reuse the mean info (instead of relying on some magic CSE?)

**Comment by rdspring1** [comment]
Ideally, we want to use the Welford one-pass algorithm, so there will only be 1 reduction. 
We'd like to make `var_mean` a prim in that case.

Reusing the mean would still require 2 reductions.

**Comment by ngimel** [comment]
What's wrong with CSE? Suppose we even could do something with this decomp, but the way users are going to write their own custom norms will be `mean=x.mean(...); var=x.var(...); now_do_something`, and we want those custom norms to have good perf. 
For layer_norm and similar cases that do use `var_mean`, nvfuser could opt out of using  var_mean decomposition and instead treat `var_mean` as a primitive, or have an fx pass that would do that? I'm reluctant to make var_mean a primitive in the general case just because nvfuser wants it.

**Comment by ezyang** [comment]
I guess my experience with CSE as an optimization is that as a program writer I don't assume it always triggers, since things like mutation can impede CSE, and unrestricted CSE can increase resident memory (since intermediates that were previously discarded are now held onto for the subsequence use.) That being said, I searched through the bug tracker for bugs for TS'es CSE pass and it seems like this never actually was a problem in practice, so someone should just write the FX CSE if it doesn't exist already (maybe @SherlockNoMad knows if there's one lying around)

**Comment by csarofeen** [comment]
Is the point here that there's already a `var_mean` prim?

**Comment by ngimel** [comment]
As a general rule, just because a backend hates a particular decomposition and wants a higher level op, that higher level op should not automatically be designated as prim. We should have a system that allows a backend to skip particular decomposition and see this operation directly - that always has been a design goal. AOTAutograd already allows one to send in a list of decompositions to use, so we can use that mechanism to avoid going into var_mean decomposition.

**Comment by csarofeen** [comment]
var_mean is super easy to decompose, and at least not trivially reconstructable. If we believe we need to have a prim to define var_mean, why don't we implement that prim instead of talking about CSE passes?

**Comment by ngimel** [comment]
We are not talking about CSE passes exclusively, we are talking about not tracing through `var_mean` decomposition, and instead NVFuser getting `var_mean` node directly (which is an alternative to CSE pass). This has been a common question from other backends such as eerie, sometimes backends want to see higher level op rather than decomposition, and we should allow this flexibility.
We are talking about CSE passes in addition because even though we can control how we write decompositions for existing operations, asking users to write `var_mean` instead of `var` and `mean` is a losing battle, so our backends should be capable of handling this user-written code too.

**Comment by csarofeen** [comment]
So this what we're discussing doing is:

1. var_mean is a reference, we should use it in the decompositions rules in norms
2. nvFuser elects to not further decompose this
3. At some point someone will make a peep hole optimization to convert separate mean/var computations (with matching dims/input) into var_mean.

is that correct?

**Comment by ngimel** [comment]
Yes. 
Also, for most layer norms welford/non-welford discussion is largely theoretical, last time I looked at nvfuser-generated code for smaller reductions that can be handled persistently (like layer norms with 1k-4k or possibly even larger sizes) NVFuser is not using welford, it's using "2-pass" mean, (x-mean)^2" approach which is not really 2-pass due to persistence, so the only real thing to do is remove second `mean` computation.

**Comment by csarofeen** [comment]
The only place we go to the 2-pass is in within block persistent normalizations. Grid normalizations need Welford, and of course non-persistent normalizations are using welford.

**Comment by ngimel** [comment]
Most (all?) layer norms are block persistent.

**Comment by csarofeen** [comment]
Sure, but if we start with welford and decompose it further, then we don't need CSE for normalizations we decompose.

But yeah, sorry, we would need both a CSE and a peep-hole optimization to reconstruct mean then var to a var_mean in user written ops.

**Comment by SherlockNoMad** [comment]
> I guess my experience with CSE as an optimization is that as a program writer I don't assume it always triggers, since things like mutation can impede CSE, and unrestricted CSE can increase resident memory (since intermediates that were previously discarded are now held onto for the subsequence use.) That being said, I searched through the bug tracker for bugs for TS'es CSE pass and it seems like this never actually was a problem in practice, so someone should just write the FX CSE if it doesn't exist already (maybe @SherlockNoMad knows if there's one lying around)

@yushangdi has recently implemented a CSE, and I hope to have it generalizes as a IR-agnostic CSE.

**Comment by jjsjann123** [comment]
> we are talking about not tracing through `var_mean` decomposition, and instead NVFuser getting `var_mean` node directly

^^^ Does this mean that the IR lowered to nvfuser would be a combination of prim & ref ops? Not a big fan of that....

**Comment by jjsjann123** [comment]
> I don't see it as such an issue. It seems we should generally stop thinking about `refs` vs `prims`. They're all just operations, and nvFuser needs to be able to specify which nodes it considers `prims`. This reminds me of what @IvanYashchuk was saying recently about how we select references that can be decomposed. For now it seems we'll have to manually select which refs we want to be able to handle. It might be nice in the future to have a programmatic mechanism to push "up" the `ref` decomposition pipe, so we can specify which `refs/prims` we support and have a programmatic way to figure out how to far to decompose for us. It seems like odd hoops to jump through, but it is what it is and I don't see a good point in litigating this further.

I agree that for the purpose of lowering from fx to nvfuser, we don't care if a node belongs to `refs` or `prims`. And I agree it should be an option to backend on how certain ops should be decomposed to.

I'm worried that with a mixed fx graph containing `refs` and `prims`, we lose some of the explicit representation of primtorch. e.g. explicit broadcast, type promotion for today. More over, primtorch infra like shape propagation would only work on a pure prim graph, IIUC.

**Comment by ngimel** [comment]
We are, hopefully, talking about a handful of ops, shape propagation can still use decomposed version if we need it (or we can have special shape prop for that handful of ops), explicit broadcast is not an issue for most ops in the list, type promotion kinda is, but again, that handful of ops is doing very limited type promotion, so it's nbd.
